src/store/hdap.js

```diff
@@ -0,0 +1,402 @@
+import { defineStore } from 'pinia'
+import { ref } from 'vue'
+
+export const useHdapStore = defineStore('hdap', () => {
+    /*******************************
+     * Internal fields and functions
+     *******************************/
+
+    async function action(name, id, optionalParams, body = {}) {
+        if (!name) return Promise.reject(new Error('You must set the action name'))
+        const bodyString = JSON.stringify(body)
+        const localParams = optionalParams || {}
+        localParams['_action'] = name
+        return fetch(new Request(`${apiBase.value}/${id}${getOptionalQueryString(localParams)}`, {
+            method: 'POST',
+            headers: getCommonHeaders(bodyString),
+            body: bodyString
+        }))
+            .then(response => returnJson(response))
+    }
+
+    function getCommonHeaders(bodyString = null) {
+        const headers = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+        if (isJwtValid()) headers['Authorization'] = `Bearer ${jwt.value}`
+        if (bodyString) headers['Content-Length'] = (new TextEncoder().encode(bodyString)).length
+        return headers
+    }
+
+    function getDateString(secondsFromNow) {
+        const future = Number(Math.floor(new Date().getTime() / 1000)) + Number(secondsFromNow)
+        return new Date(future * 1000).toISOString()
+    }
+
+    function getExpiredDateString() {
+        return new Date(0).toISOString() // '1970-01-01T00:00:00.000Z'
+    }
+
+    function getOptionalQueryString(params) {
+        return (params) ? '?' + new URLSearchParams(params) : ''
+    }
+
+    function isJwtValid() {
+        return new Date(expires.value).getTime() > new Date().getTime()
+    }
+
+    async function put(id, resource, additionalHeaders, optionalParams) {
+        if (!id || !resource) throw new Error('You must set the resource identifier and the resource')
+        const bodyString = JSON.stringify(resource)
+        const headers = getCommonHeaders(bodyString)
+        for (const [key, value] of Object.entries(additionalHeaders)) {
+            headers.headers[key] = value
+        }
+        return fetch(new Request(`${apiBase.value}/${id}${getOptionalQueryString(optionalParams)}`, {
+            method: 'PUT',
+            headers: headers,
+            body: bodyString
+        }))
+            .then(response => returnJson(response))
+    }
+
+    async function returnJson(response) {
+        if (!response.ok) throw new Error(`HTTP error with status: ${response.status}`)
+        return response.json()
+    }
+
+    async function throwError(message) {
+        message.value = message
+        throw new Error(message)
+    }
+
+    /*******************************
+     * Exported fields and functions
+     *******************************/
+
+    /** Base path to HDAP. */
+    const apiBase = ref('/hdap')
```

Comment:
I guess this should be imported somehow from https://github.com/markcraig/hdap-demo/blob/main/vite.config.mjs#L66